File: prismpath/examples/cli_worker/ci_gate.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""CI gate worker (Python). Read the build report from the [context] block on stdin, decide pass/coverage,
print ONE JSON object, exit 0. A nonzero exit routes to the flow's error tier.
Wire it in with:  cli_agent(["python", "ci_gate.py"], pass_state=["report"])"""
import json
import re
import sys


def context() -> dict:
    _, _, ctx = sys.stdin.read().partition("[context]")   # PrismPath appends: ...\n\n[context]\n{json}
    return json.loads(ctx.strip() or "{}")


try:
    report = context()["report"]
    failed = int(re.search(r"failed=(\d+)", report).group(1))
    coverage = int(re.search(r"coverage=(\d+)", report).group(1))
except (KeyError, AttributeError, ValueError):
    print("unparseable build report", file=sys.stderr)     # -> the flow's error tier
    sys.exit(1)

print(json.dumps({"passed": failed == 0, "failed": failed, "coverage": coverage}))

File: prismpath/examples/cli_worker/ci_gate.rs
// CI gate worker (Rust) — the same job as ci_gate.py, to show one job porting across languages with no
// change to the flow. Read the build report from stdin, decide pass/coverage, print ONE JSON object,
// exit 0. A nonzero exit routes to the flow's error tier. This scans stdin for the report's fields; a
// worker needing structured input would parse the [context] JSON block (see ci_gate.py).
// Build once (rustc -O ci_gate.rs), then wire it in with:  cli_agent(["./ci_gate"], pass_state=["report"])
use std::io::{self, Read};

fn field(text: &str, key: &str) -> Option<i64> {
    let at = text.find(key)?;
    let digits: String = text[at + key.len()..]
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect();
    digits.parse().ok()
}

fn main() {
    let mut text = String::new();
    io::stdin().read_to_string(&mut text).unwrap();
    match (field(&text, "failed="), field(&text, "coverage=")) {
        (Some(failed), Some(coverage)) => print!(
            "{{\"passed\":{},\"failed\":{},\"coverage\":{}}}",
            failed == 0,
            failed,
            coverage
        ),
        _ => {
            eprint!("unparseable build report"); // -> the flow's error tier
            std::process::exit(1);
        }
    }
}
